src/Transports.AspNetCore/WebSockets/SubscriptionList.cs

```diff
@@ -0,0 +1,228 @@
+namespace GraphQL.Server.Transports.AspNetCore.WebSockets;
+
+/// <summary>
+/// A list of subscriptions represented by <see cref="IDisposable"/> objects.
+/// All members are thread-safe.
+/// <br/><br/>
+/// Upon signalling the cancellation token, all future calls to any method except
+/// <see cref="Dispose"/> will throw an <see cref="OperationCanceledException"/>.
+/// <br/><br/>
+/// It is recommended to call <see cref="Dispose"/> immediately after signalling
+/// the cancellation token.
+/// </summary>
+public sealed class SubscriptionList : IDisposable
+{
+    private readonly CancellationToken _cancellationToken;
+    private readonly Dictionary<string, IDisposable> _subscriptions = new();
+    private readonly object _lock = new();
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    public SubscriptionList(CancellationToken cancellationToken)
+    {
+        _cancellationToken = cancellationToken;
+    }
+
+    /// <summary>
+    /// Disposes of all active subscriptions.
+    /// </summary>
+    public void Dispose()
+    {
+        List<IDisposable> subscriptionsToDispose;
+        lock (_lock)
+        {
+            subscriptionsToDispose = _subscriptions.Values.ToList();
+            _subscriptions.Clear();
+        }
+        foreach (var disposer in subscriptionsToDispose)
+        {
+            disposer.Dispose();
+        }
+    }
+
+    /// <summary>
+    /// Adds a subscription to the internal list if none is already in the list with the same id.
+    /// </summary>
+    /// <exception cref="ArgumentNullException"/>
+    /// <exception cref="OperationCanceledException"/>
+    public bool TryAdd(string id, IDisposable subscription)
```

Comment:
I would write a note about _id_ argument mentioning that this is the subscription identifier or something.

Reply:
I altered the class comment.